java-checks/src/main/java/org/sonar/java/checks/AssertionsInTestsCheck.java

```diff
@@ -0,0 +1,121 @@
+  private static final MethodInvocationMatcher FEST_DESCRIBED_AS_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(GENERIC_ASSERT).name("describedAs").withNoParameterConstraint();
+  private static final MethodInvocationMatcher FEST_OVERRIDE_ERROR_METHOD = MethodInvocationMatcher.create()
+    .typeDefinition(GENERIC_ASSERT).name("overridingErrorMessage").withNoParameterConstraint();
+  private static final ArrayList<MethodInvocationMatcher> ASSERTION_INVOCATION_MATCHERS = Lists.newArrayList(
+    MethodInvocationMatcher.create().typeDefinition("org.junit.Assert").name(NameCriteria.startsWith("assert")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("org.junit.Assert").name("fail").withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("junit.framework.Assert").name(NameCriteria.startsWith("assert")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("junit.framework.Assert").name(NameCriteria.startsWith("fail")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition("org.fest.assertions.Fail").name(NameCriteria.startsWith("fail")).withNoParameterConstraint(),
+    MethodInvocationMatcher.create().typeDefinition(TypeCriteria.subtypeOf(GENERIC_ASSERT)).name(NameCriteria.any()).withNoParameterConstraint()
+  );
+
+  private Deque<Boolean> methodContainsAssertion = new ArrayDeque<>();
+  private Deque<Boolean> inUnitTest = new ArrayDeque<>();
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    boolean isUnitTest = isUnitTest(tree);
+    inUnitTest.push(isUnitTest);
+    methodContainsAssertion.push(Boolean.FALSE);
+    super.visitMethod(tree);
+    inUnitTest.pop();
+    if (isUnitTest && !methodContainsAssertion.pop()) {
+      context.addIssue(tree, this, "Add at least one assertion to this test case.");
+    }
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    super.visitMethodInvocation(tree);
+    if (inUnitTest.peek() && !methodContainsAssertion.peek() && isAssertion(tree)) {
+      methodContainsAssertion.pop();
+      methodContainsAssertion.push(Boolean.TRUE);
+    }
+  }
+
+  private boolean isAssertion(MethodInvocationTree mit) {
+    for (MethodInvocationMatcher methodInvocationMatcher : ASSERTION_INVOCATION_MATCHERS) {
+      if (methodInvocationMatcher.matches(mit) &&
+          !FEST_AS_METHOD.matches(mit) &&
+          !FEST_OVERRIDE_ERROR_METHOD.matches(mit) &&
+          !FEST_DESCRIBED_AS_METHOD.matches(mit)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isUnitTest(MethodTree tree) {
+    for (AnnotationTree annotationTree : tree.modifiers().annotations()) {
```

Comment:
It is fine to do it via syntax but (so you know) you can also directly rely on semantic and use the metadata of the method symbol to check annotation. (you might end up with less code).